bokeh/document.py

```diff
@@ -25,6 +26,38 @@
 
 DEFAULT_TITLE = "Bokeh Application"
 
+class DocumentProxy(object):
+    ''' Wrap a Document object so that only methods that can safely be used
+    from unlocked callbacks or threads are exposed. Attempts to otherwise
+    access or change the Document results in an exception.
+
+    '''
+
+    def __init__(self, doc):
+        self._doc = doc
+
+    def __getattr__(self, attr):
+        raise RuntimeError("Only add_next_tick_callback may be used safely from DocumentProxy")
```

Comment:
Yah, I realized that as well, but that also generates an exception... maybe we could make it better/more verbose, but we can also certainly document loudly around `@without_document_lock` and since it is a thing that has to be done affirmatively and explicitly, I think it's ok. I will make the suggested other changes, your messages are always great :)